### PR TITLE
[BREAKING] solidity-upgrade: added module to remove visibility of constructors

### PR DESCRIPTION
**Pull request by hrkrshnn**

### CODE REVIEW

File: tools/solidityUpgrade/SourceTransform.h [hunk 1]
```diff
@@ -274,6 +274,15 @@ class SourceTransform
 	{
 		return regex_replace(_location.text(), std::regex{"now"}, "block.timestamp");
 	}
+
+	static std::string removeVisibility(langutil::SourceLocation const& _location)
+	{
+		std::string replacement = _location.text();
+		std::vector toReplace{"public ", "public", "internal ", "internal"};
+		for (auto const& replace: toReplace)
```
Comment (Marenz) [inline tools/solidityUpgrade/SourceTransform.h]:
```suggestion
		for (auto const& replace: {"public ", "public", "internal ", "internal"})
```

**Review by Marenz — CHANGES_REQUESTED** [review]
LGTM, of course test are failing and one suggestion to write something shorter

**Comment by hrkrshnn** [comment]
The zeppelin external contracts needs fixing. I believe, the CI is using solidity-nightly to run the tests instead of 0.7.0 (whatever solc created during tests.)

File: tools/solidityUpgrade/Upgrade070.cpp [hunk 2]
```diff
@@ -59,3 +59,16 @@ void NowKeyword::endVisit(Identifier const& _identifier)
 			);
 		}
 }
+
+void ConstructorVisibility::endVisit(ContractDefinition const& _contract)
+{
+	// TODO If constructor is internal, maybe it makes sense to change contract to abstract. But
```
Comment (chriseth) [inline tools/solidityUpgrade/Upgrade070.cpp]: Yes, it should be made abstract in this case.

**Comment by axic** [comment]
#9180 was merged, this needs to be rebased.

File: tools/solidityUpgrade/SourceTransform.h [hunk 3]
```diff
@@ -132,7 +132,8 @@ class SourceTransform
 			return regex_replace(
 				_location.text(),
 				_regex,
-				_expression + " " + _keyword
+				_expression + " " + _keyword,
+				std::regex_constants::format_first_only
```
Comment (hrkrshnn) [inline tools/solidityUpgrade/SourceTransform.h:136]: Without this flag, the following would be upgraded to 

```solidity
contract test {
	/// contract upgrade
	constructor() internal {}
}
```

```solidity
abstract contract test {
	/// abstract contract upgrade
	constructor() {}
}
```

File: docs/using-the-compiler.rst [hunk 4]
```diff
@@ -635,116 +635,88 @@ If you found a bug or if you have a feature request, please
...
 
+
 Required changes
 ^^^^^^^^^^^^^^^^
 
-To bring the contracts up to date with the current Solidity version, the
-following upgrade modules have to be executed: ``constructor``,
-``visibility``, ``abstract``, ``override`` and ``virtual``. Please read the
-documentation on :ref:`available modules <upgrade-modules>` for further details.
+The above contract will not compile after 0.7.0. To bring the contract up to date with the current
```
Comment (Marenz) [inline docs/using-the-compiler.rst]: The wording is a bit confusing "after 0.7.0" makes me think of "0.8.0". I suggest using "with" or "starting with".

"will no longer compile starting with 0.7.0"

**Review by Marenz — CHANGES_REQUESTED** [review]

File: docs/using-the-compiler.rst (hunk 4 above)
Comment (Marenz) [reply docs/using-the-compiler.rst]: Any change to this?

**Review by Marenz — APPROVED** [review]
LGTM, only the ch_proof test is failing

**Pull request merged**